references/cli/addon.go

```diff
@@ -428,6 +428,70 @@ If you don't want the Helm component, just omit the three Chart-related paramete
 	return cmd
 }
 
+// NewAddonPushCommand pushes an addon dir/package to a ChartMuseum
+func NewAddonPushCommand(c common.Args) *cobra.Command {
+	p := &pkgaddon.PushCmd{}
+	cmd := &cobra.Command{
+		Use:   "push",
+		Short: "uploads an addon package to ChartMuseum",
+		Long: `Uploads an addon package to ChartMuseum.
+
+Two arguments are needed <addon directory/package> and <name/URL of ChartMuseum>.
+
+The first argument <addon directory/package> can be:
+	- your conventional addon directory (containing metadata.yaml). We will package it for you.
+	- packaged addon (.tgz) generated by 'vela addon package' command
+
+The second argument <name/URL of ChartMuseum> can be:
+	- registry name (helm type). You can add your ChartMuseum registry using 'vela addon registry add'.
+	- ChartMuseum URL, e.g. http://localhost:8080`,
+		Example: `# Push the addon in directory <your-addon> to a ChartMuseum registry named <localcm>
+$ vela addon push your-addon localcm
+
+# Push packaged addon mongo-1.0.0.tgz to a ChartMuseum registry at http://localhost:8080
+$ vela addon push mongo-1.0.0.tgz http://localhost:8080
+
+# Force push, overwriting existing ones
+$ vela addon push your-addon localcm -f`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 2 {
+				return fmt.Errorf("two arguments are needed: addon directory/package, name/URL of Chart repository")
+			}
+
+			c, err := c.GetClient()
+			if err != nil {
+				return err
+			}
+
+			p.Client = c
+			p.Out = cmd.OutOrStdout()
+			p.ChartName = args[0]
+			p.RepoName = args[1]
+			p.SetFieldsFromEnv()
```

Comment:
is there another way other than setting the environment and use flags? What are the ways for helm？

Reply:
I think helm has two ways, 1. environment vars, 2. command line arguments. We implemented these two as well.

Reply:
can we re-use their config if users already have one? read the same defualt path is even better

Reply:
I don't think that config file is appropriate here (so I didn't count it). That config file is for `Codefresh CLI`, and the repo need to be inside `helm`. It stores username and password (tokens, to be precise), so the user don't have to input their credentials every time they push anything.

We are taking advantage of `addon registries`, so the same feature is available. If the user add a basic-auth authenticated registry to `addon registries`, they don't need to input their credentials when they `vela addon push`. Because I will read the credentials from `addon registries`.

Reply:
great, that's I want, but I didn't see the logic in your code.

> If the user add a basic-auth authenticated registry to addon registries, they don't need to input their credentials when they vela addon push. Because I will read the credentials from addon registries.

Reply:
It's here
https://github.com/kubevela/kubevela/blob/229ea1999fe9c20f09910d216c38e8d5685f929f/pkg/addon/push.go#L210

Reply:
BUt you're not reading the credentials for certificates

Reply:
custom certificates are not supported by `addon registries` anyway...

Reply:
anyway, you pointed the right direction we should provide